### PR TITLE
Fix removal race condition in ListContainers

### LINKED ISSUE #10120: Listing containers fails with "container xxx does not exist in database: no such container"
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**
I encountered the same bug as described in #418 using the podman API

>could not list containers: Error response from daemon: container bf4677025fecfa5c4f0883537d857d76a613a5ca7c7c9d79b81c1b7ec2dc277a does not exist in database: no such container

I'm using the go docker API library and get the error during tests that create and delete containers concurrently.

**Steps to reproduce the issue:**

This test reliably provokes the error
```go
package main

import (
	"context"
	"strconv"
	"testing"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"
)

func Test_ProvokeErr(t *testing.T) {
	podClient, err := client.NewEnvClient()
	if err != nil {
		t.Fatal(err)
	}
	for i := 0; i < 50; i++ {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			t.Parallel()
			creationResp, err := podClient.ContainerCreate(
				context.Background(),
				&container.Config{Image: "alpine"},
				nil, nil, "",
			)
			if err != nil {
				t.Fatal(err)
			}

			_, err = podClient.ContainerList(context.Background(), types.ContainerListOptions{All: true})
			if err != nil {
				t.Error("could not list containers:", err)
			}

			err = podClient.ContainerRemove(context.Background(), creationResp.ID, types.ContainerRemoveOptions{})
			if err != nil {
				t.Error(err)
			}
		})
	}
}
```

**Describe the results you received:**
```log
test_test.go:32: could not list containers: Error response from daemon: container edb1f86b964f08558aa2ba4799aa81db59c2fe3ff9754c704252db2df708682e does not exist in database: no such container
test_test.go:32: could not list containers: Error response from daemon: container c0114fc1f0a10a809db3a681fbd0d612aa9e87b4170a08f5a9b028ec442e33c1 does not exist in database: no such container
...
```
**Describe the results you expected:**
no errors

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:      3.1.0
API Version:  3.1.0
Go Version:   go1.16
Built:        Tue Mar 30 15:29:36 2021
OS/Arch:      linux/amd64
```

<details>
  <summary>Output of `podman info --debug`</summary>

```
host:
  arch: amd64
  buildahVersion: 1.20.0
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.27-1.fc34.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.27, commit: '
  cpus: 16
  distribution:
    distribution: fedora
    version: "34"
  eventLogger: journald
  hostname: honestmistake
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.11.12-300.fc34.x86_64
  linkmode: dynamic
  memFree: 6679322624
  memTotal: 32910741504
  ociRuntime:
    name: crun
    package: crun-0.18-5.fc34.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.18
      commit: 808420efe3dc2b44d6db9f1a3fac8361dde42a95
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    selinuxEnabled: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.9-1.fc34.x86_64
    version: |-
      slirp4netns version 1.1.8+dev
      commit: 6dc0186e020232ae1a6fcc1f7afbc3ea02fd3876
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.0
  swapFree: 8589930496
  swapTotal: 8589930496
  uptime: 4h 49m 33.47s (Approximately 0.17 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /home/jakob/.config/containers/storage.conf
  containerStore:
    number: 4
    paused: 0
    running: 3
    stopped: 1
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.5.0-1.fc34.x86_64
      Version: |-
        fusermount3 version: 3.10.2
        fuse-overlayfs: version 1.5
        FUSE library version 3.10.2
        using FUSE kernel interface version 7.31
  graphRoot: /home/jakob/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 42
  runRoot: /run/user/1000/containers
  volumePath: /home/jakob/.local/share/containers/storage/volumes
version:
  APIVersion: 3.1.0
  Built: 1617110976
  BuiltTime: Tue Mar 30 15:29:36 2021
  GitCommit: ""
  GoVersion: go1.16
  OsArch: linux/amd64
  Version: 3.1.0
```

</details>

<details>
  <summary>Package info</summary>

```
Name         : podman
Epoch        : 2
Version      : 3.1.0
Release      : 1.fc34
Architecture : x86_64
Size         : 46 M
Source       : podman-3.1.0-1.fc34.src.rpm
Repository   : @System
From repo    : fedora
Summary      : Manage Pods, Containers and Container Images
URL          : https://podman.io/
License      : ASL 2.0
Description  : podman (Pod Manager) is a fully featured container engine that is a simple
             : daemonless tool.  podman provides a Docker-CLI comparable command line that
             : eases the transition from other container engines and allows the management of
             : pods, containers and images.  Simply put: alias docker=podman.
             : Most podman commands can be run as a regular user, without requiring
             : additional privileges.
             : 
             : podman uses Buildah(1) internally to create container images.
             : Both tools share image (not container) storage, hence each can use or
             : manipulate images (but not containers) created by the other.
             : 
             : Manage Pods, Containers and Container Images
             : podman Simple management tool for pods, containers and images
```
</details>

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

### PR DESCRIPTION
**Pull request by SoMuchForSubtlety**
It is possible that a container is removed between fetching the initial list of containers and the second access during conversion. In that case it can just be dropped from the returned list.

A similar issue was previously fixed with 3fbc2c5cd21a344819a0636f08487cee805dd8d9

We might want to change from preallocating the slice of containers to just appending because adding an offset makes the code less readable.

I tried to add a test to `pkg/bindings/test/containers_test.go`, but creating and deleting containers in parallel caused unrelated errors. The test I provided in #10120  passes with this fix.

Closes #10120

### CODE REVIEW

File: pkg/api/handlers/compat/containers.go [hunk 1]
```diff
@@ -148,16 +149,23 @@ func ListContainers(w http.ResponseWriter, r *http.Request) {
 			containers = containers[:query.Limit]
 		}
 	}
-	var list = make([]*handlers.Container, len(containers))
+	list := make([]*handlers.Container, len(containers))
```
Comment (vrothberg) [inline pkg/api/handlers/compat/containers.go:152]: Can we turn this into `var list []*handlers.Container` and use `append = ....` in the loop?  It'll be slightly less efficient but more robust.

**Review by vrothberg — COMMENTED** [review]
Thanks!

Can you add `[NO TESTS NEEDED]` to the commit message?  We enforce adding tests with each PR but in this case I think we don't need some (races are hard to test).

File: pkg/api/handlers/compat/containers.go (hunk 1 above)
Comment (SoMuchForSubtlety) [reply pkg/api/handlers/compat/containers.go]: Now golangci-lint complains, should I add `//nolint`?

Comment (vrothberg) [reply pkg/api/handlers/compat/containers.go]: Does it also complain when turning it into `list := []*handlers.Container{}`?  You can run it locally with `make lint`.

Comment (SoMuchForSubtlety) [reply pkg/api/handlers/compat/containers.go]: That works.

Comment (Luap99) [reply pkg/api/handlers/compat/containers.go]: The best way is to preallocate with `make([]*handlers.Container, 0, len(containers))`. This allows you to use append correctly.

Comment (vrothberg) [reply pkg/api/handlers/compat/containers.go]: Great idea, @Luap99 !

**Review by vrothberg — COMMENTED** [review]
LGTM

**Review by Luap99 — APPROVED** [review]
LGTM

**Pull request merged**